### ISSUE TITLE
`code` CLI still opens duplicate VSCode instance in the macOS dock

### ISSUE THREAD

**Issue by lgarron** [microsoft/vscode#139634]
Issue Type: <b>Bug</b>
This is a re-file of https://github.com/microsoft/vscode/issues/60579, which does not appear to be fixed for me in the November VSCode release.

VSCode is still opening processes with duplicate dock icons (and therefore slowing down the use of `code` to open files). Judging from the issue and the accompanying PR, it looks like I'm not the only one who is still experiencing this issue. Unfortunately, the other issues and PRs are closed and locked, so I'm filing this new one.

Simple repro: `code new-file.txt`

One workaround from the other issue was to use `open -b com.microsoft.VSCode`. However, this has some challenges:
- The `open` command completely fails on macOS if passed any file that does not exist yet. This makes the raw command useless for creating new files from the commandline. 😔
  - `open -b com.microsoft.VSCode new-file` outputs: `The file /Users/lgarron/new-file does not exist.`
- It is not obvious how to craft a shell function wrapper that properly opens files, can handle multiple paths (files or folders), and isn't subject to edge cases and injections (e.g. file names that will be confused for arguments to `open`).
  - `open -b com.microsoft.VSCode --args new-file` does nothing, outputs nothing, and exits without an error. Leaving out `--args` is unsafe (when the file name is not typed by hand and known not to conflict with an argument to `open`) but doesn't work at all.
- Using `open` a wrapper for as a workaround for `code` breaks non-file arguments — even `code --help`.
- In addition to setting an alias/function for `code` in your shell, you will need to set up `$EDITOR` separately (at least in some shells).

A proper fix would be really appreciated.

VS Code version: Code 1.63.2 (899d46d82c4c95423fb7e10e68eba52050e30ba3, 2021-12-15T09:37:28.172Z)
OS version: Darwin x64 21.2.0
Restricted Mode: No

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i9-8950HK CPU @ 2.90GHz (12 x 2900)|
|GPU Status|2d_canvas: enabled<br>gpu_compositing: enabled<br>metal: disabled_off<br>multiple_raster_threads: enabled_on<br>oop_rasterization: enabled<br>opengl: enabled_on<br>rasterization: enabled<br>skia_renderer: disabled_off_ok<br>video_decode: enabled<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|4, 3, 3|
|Memory (System)|32.00GB (0.14GB free)|
|Process Argv|--crash-reporter-id 98789cd9-b9d1-4ac0-b09a-99c0a9d11695|
|Screen Reader|no|
|VM|0%|
</details><details><summary>Extensions (12)</summary>

Extension|Author (truncated)|Version
---|---|---
custom-window-zoom|cra|1.1.4
vscode-eslint|dba|2.2.2
prettier-vscode|esb|9.0.0
codespaces|Git|1.4.2
vscode-pull-request-github|Git|0.34.2
go|gol|0.30.0
jupyter-keymap|ms-|1.0.0
vsliveshare|ms-|1.0.5200
vs-code-prettier-eslint|rve|3.0.4
fish-vscode|sky|0.2.1
shellcheck|tim|0.18.3
vscode-status-bar-format-toggle|tom|3.0.0

</details><details>
<summary>A/B Experiments</summary>

```
vsliv368cf:30146710
vsreu685:30147344
python383cf:30185419
vspor879:30202332
vspor708:30202333
vspor363:30204092
vswsl492cf:30256860
pythontb:30283811
pythonvspyt551:30345470
pythonptprofiler:30281270
vshan820:30294714
vstes263cf:30335440
vscoreces:30384385
pythondataviewer:30285071
vscod805:30301674
pythonvspyt200:30340761
binariesv615:30325510
bridge0708:30335490
bridge0723:30353136
vsaa593:30376534
pythonvs932:30410667
vscop804:30404766
vscop341:30404997
vsrem710cf:30416617
vscexrecpromp3t2:30407763

```

</details>

**Comment by IllusionMH** [comment]
Do you see same issue if you disable all extensions? May be duplicate of #138712

**Comment by lgarron** [comment]
> Do you see same issue if you disable all extensions? May be duplicate of #138712

Yes, `code --disable-extensions` successfully disables extensions but does not address this issue. 😢

**Comment by lgarron** [comment]
I also just realized that opening the editor for things like `git commit` was super slow. `set EDITOR "open -b com.microsoft.VSCode"` has helped for that, but I'm worried it will break in other situations.

**Comment by lgarron** [comment]
> I just added `set EDITOR "open -b com.microsoft.VSCode"` and that's a _lot_ faster, but I'm worried it will break in other situations.

... aaaaand:

<img width="280" alt="Screen Shot 2022-01-23 at 20 58 52" src="https://user-images.githubusercontent.com/248078/150724563-06296adf-883f-4795-b7a5-ffde54244b4a.png">

😔

**Comment by hazarek** [comment]
All extensions are disabled and the new icon is still bouncing when a file is opened with the terminal in the same workspace. disappears in a few seconds

**Comment by Drarig29** [comment]
I had the same problem and I use the "Extension Bisect" feature to find the culprit.

In my case it was the extension "Spell Right". I'm posting an issue on their repo.

**Comment by TylerADavis** [comment]
What would be the best way to troubleshoot this? I just started seeing this behavior (double dock icons) the last few weeks, perhaps after an update to an extension or vs code itself. I tried `code --disable-extensions .` as in https://github.com/microsoft/vscode/issues/139634#issuecomment-999541635, but the issue persists.


![image](https://user-images.githubusercontent.com/16804460/167279099-c107e1f5-116b-4b90-ba44-acbb79c967f9.png)

**Comment by Drarig29** [comment]
> What would be the best way to troubleshoot this? I just started seeing this behavior (double dock icons) the last few weeks, perhaps after an update to an extension or vs code itself. I tried `code --disable-extensions .` as in https://github.com/microsoft/vscode/issues/139634#issuecomment-999541635, but the issue persists.
> 
> 
> ![image](https://user-images.githubusercontent.com/16804460/167279099-c107e1f5-116b-4b90-ba44-acbb79c967f9.png)
> 

As I said just before, try using the "Extension Bisect" feature. You can start it in the Extensions panel, with the "..." menu.

More information: https://code.visualstudio.com/blogs/2021/02/16/extension-bisect

**Comment by TylerADavis** [comment]
Thanks @Drarig29 ! I had expected `--disable-extensions` to stop all extensions from running and rule out the possibility of being extension related, but that seems to not be the case. I ended up having the same extension as the culprit.

**Comment by lgarron** [comment]
> * In addition to setting an alias/function for `code` in your shell, you will need to set up `$EDITOR` separately (at least in some shells).

I've tried to debug this a bunch of times, and concluded that it's impractical (or maybe even impossible) to maintain enough workarounds for this. I use `fish`, and all the combinations of abbreviations / `$EDITOR` / `$VISUAL` just don't seem to work for the "edit commandline with external editor "functionality.

If I want to use the workaround for the slow duplicate VSCode instance (which still happens for me), I inevitably have to leave other things broken right now. 😕

**Comment by lgarron** [comment]
VSCode has been frustratingly slow to load[^1] for me, and I've been having issues pinning it down, since it occurs even for `code --disable-extensions`. I now believe the root cause is related to this issue. :-/

[^1] Slow to load, slow to create files, slow to open files. Once a file is open, things are fast, though.

**Comment by pcutillas** [comment]
This isn't a direct solution to this, but if you don't use the recent apps section of the dock, you can disable it: https://www.howtogeek.com/714379/how-to-hide-recent-apps-from-the-mac-dock/#:~:text=In%20System%20Preferences%2C%20select%20%E2%80%9CDock,will%20disappear%20from%20your%20Dock.

**Comment by lgarron** [comment]
> This isn't a direct solution to this, but if you don't use the recent apps section of the dock, you can disable it: https://www.howtogeek.com/714379/how-to-hide-recent-apps-from-the-mac-dock/#:~:text=In%20System%20Preferences%2C%20select%20%E2%80%9CDock,will%20disappear%20from%20your%20Dock.

For what it's worth, I've personally had the recent apps section disabled for a long time, and it seems to be unrelated to this issue (which is still happening, and I've just learned to live with the slight slowdown and extra visual distraction of using the original `code` binary as my editor so that I don't have to deal with the bugs and security issues of the workaround alternatives).

**Comment by plttn** [comment]
I can replicate this with zero extensions enabled, as long as Code is running. 

In other words: 
1. run `code .` from Terminal
2. Only pinned Dock icon bounces.
3. work on that window for a bit, then realize "oh I need to work on another project"
4. use the Terminal to navigate to the new project
5. run `code .` 
6. duplicate icon appears

**Comment by conradwt** [comment]
@plttn I'm starting seeing a similar issue.  Furthermore, the more times I execute `code-insiders .` the more VSCode icons I'll see in the Dock.  This issue always start with the editor window not resizing to the full size of the monitor.

**Comment by cameronelliott** [comment]
Also seeing this issue.
<img width="817" alt="image" src="https://github.com/microsoft/vscode/assets/868689/330c4870-8993-413c-85ec-162d4e67b1bb">

**Comment by thadbloom** [comment]
I solved it on my machine with this:

`alias code="open -b com.microsoft.VSCode"`

Not really an actual solution. But this was annoying to me.

**Comment by conradwt** [comment]
Here's a recent error report and a picture of what I'm seeing on the screen when I attempt to make the image full-screen.

error report:

https://gist.github.com/conradwt/4fcdc1f3447ea3306923cdccf7e98191

![Screenshot 2023-09-25 at 7 58 12 PM](https://github.com/microsoft/vscode/assets/20219/37b89c07-d3df-4b39-be02-ad98a1bbbd9d)

**Comment by davilima6** [comment]
I confirm the issue persists using latest. Besides, the `open` workaround fails to jump to a specific line as `--args` is ignored:

```
/usr/bin/open -b com.microsoft.VSCode --args --goto myFile.ext:10
```

Which I need for _Run Command_ on iTerm's Semantic History: `/usr/bin/open -b com.microsoft.VSCode --args --goto \1:\2`

Using:

```
Version: 1.85.1 (Universal)
Commit: 0ee08df0cf4527e40edc9aa28f4b5bd38bbff2b2
Date: 2023-12-13T09:48:06.308Z
Electron: 25.9.7
ElectronBuildId: 25551756
Chromium: 114.0.5735.289
Node.js: 18.15.0
V8: 11.4.183.29-electron.0
OS: Darwin arm64 23.2.0
```

**Comment by cameronelliott** [comment]
Yeah,the work around is lame, it broke my arguments from working also.
It would be nice to not need the work around so the regular flags worked.

**Comment by conradwt** [comment]
After upgrading to macOS Sonoma, all the VSCode - Insider issues that I reported went away for me.

**Comment by davilima6** [comment]
I'm using macOS Sonoma 14.2.1 and I confirm the issue persists using the latest regular (non-Insider) VS Code.

Besides, the `open` workaround fails to jump to a specific line as `--args` is ignored:

```
open -b com.microsoft.VSCode --args --goto myFile.ext:10
```

Which I need for _Run Command_ on iTerm's Semantic History:

```
/usr/bin/open -b com.microsoft.VSCode --args --goto \1:\2
```

Using:

```
Version: 1.85.1 (Universal)
Commit: 0ee08df0cf4527e40edc9aa28f4b5bd38bbff2b2
Date: 2023-12-13T09:48:06.308Z
Electron: 25.9.7
ElectronBuildId: 25551756
Chromium: 114.0.5735.289
Node.js: 18.15.0
V8: 11.4.183.29-electron.0
OS: Darwin arm64 23.2.0
```

**Comment by fserb** [comment]
On MacOS Sonoma 14.2.1, seeing multiple instances of VSCode when using the `code` cli.

The `open` workaround fails to parse arguments properly (you can check by trying: `open -b com.microsoft.VSCode --args --new-window` doesn't work at all.

**Comment by AntonellaMorittu** [comment]
In your Mac Dock Settings, disable the recent apps in the dock and it will work just fine 👍 

<img width="472" alt="Screenshot 2024-03-15 at 11 41 24" src="https://github.com/microsoft/vscode/assets/33493978/e23fe9ea-e9a2-40fe-8b4e-54083b5f3d60">

**Comment by lgarron** [comment]
> In your Mac Dock Settings, disable the recent apps in the dock and it will work just fine 👍
> 
> <img alt="Screenshot 2024-03-15 at 11 41 24" width="472" src="https://private-user-images.githubusercontent.com/33493978/313160735-e23fe9ea-e9a2-40fe-8b4e-54083b5f3d60.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTA1MTAyMDUsIm5iZiI6MTcxMDUwOTkwNSwicGF0aCI6Ii8zMzQ5Mzk3OC8zMTMxNjA3MzUtZTIzZmU5ZWEtZTlhMi00MGZlLThiNGUtNTQwODNiNWYzZDYwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzE1VDEzMzgyNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThkNjEzNDE4MzlmMmY5MjU1NGM3MzkzZWYwOTRhMzM0ZDA0MmE0Y2I4MzJkZDAzOWEyNWU2YjIwM2U3NmI2OTgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.9xLwsWdUZPalY-zh9VBy5VCisqL9jPgLhIjcLuc296I">

I've had recent apps disabled in the dock for a long time. It didn't help for me [a year ago](https://github.com/microsoft/vscode/issues/139634#issuecomment-1428630310) and still doesn't help now, unfortunately. The bug still persists.

**Comment by AlbertShown** [comment]
macOS... really annoying.
`code --resuse-window`, `alias code="open -b com.microsoft.VSCode"`... all doesn't work.

**Comment by OliverJAsh** [comment]
> Besides, the `open` workaround fails to jump to a specific line as `--args` is ignored:

FWIW this works: `open vscode://file/<path>:<line>`.

**Comment by easazade** [comment]
I created this function as a workaround and added it to `~/.zshrc`
```bash
function code(){
    echo "workaround for code command: open -b com.microsoft.VSCode $*"
    open -b com.microsoft.VSCode "$@"
}
```

**Comment by rzhao271** [comment]
Assigning back to myself and adding help wanted because we do not have the bandwidth to look into this issue.
Our previous attempt to fix the issue with PR https://github.com/microsoft/vscode/pull/259883 resulted in issue https://github.com/microsoft/vscode/issues/274146.

**Comment by AlbertShown** [comment]
@rzhao271 Someone needs to contact Apple and tell them to cut the crap.

**Comment by OliverJAsh** [comment]
@rzhao271 I'm curious if https://github.com/microsoft/vscode/pull/259883 also fixed the performance issue? The duplicate dock instance is one thing but worse is how slow `code` is to open the file (when compared to `open` which is instant). I'm not sure if this has the same root cause as the duplicate dock issue.
